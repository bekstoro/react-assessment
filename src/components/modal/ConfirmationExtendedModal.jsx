import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Portal from '@components/portal/Portal';
import {
  BigModal,
  ModalButtons,
  ModalCancellationButton,
  ModalConfirmationButton,
  ModalContent,
  ModalError,
  ModalInput,
  ModalTitle,
  RedIcon,
} from '@components/styled-components/Modal';
import { DELETE } from '@constants';

export default function ConfirmationExtendedModal({
  cancellationButtonText,
  confirmationButtonText,
  isOpen,
  message,
  title,
  closeModal,
  confirmModal,
}) {
  const [confirmValue, setConfirmValue] = useState('');
  const [error, setError] = useState(false);

  if (!isOpen) return null;

  return (
    <Portal closeModal={closeModal}>
      <BigModal data-testid="confirmation-extended-modal">
        <ModalTitle>
          <RedIcon className="material-icons">warning</RedIcon>
          {title}
        </ModalTitle>
        <ModalContent>
          {message}
        </ModalContent>
        <ModalInput value={confirmValue} onChange={(e) => setConfirmValue(e.target.value)} />
        {error && <ModalError>Confirmation word is mismatched</ModalError>}
        <ModalButtons>
          <ModalConfirmationButton
            onClick={() => {
              if (confirmValue === DELETE) {
                confirmModal();
                closeModal();
                setError(false);
                setConfirmValue('');
              } else {
                setError(true);
              }
            }}
          >
            {confirmationButtonText}
          </ModalConfirmationButton>
          <ModalCancellationButton
            onClick={() => {
              closeModal();
              setConfirmValue('');
            }}
          >
            {cancellationButtonText}
          </ModalCancellationButton>
        </ModalButtons>
      </BigModal>
    </Portal>
  );
}

ConfirmationExtendedModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  cancellationButtonText: PropTypes.string.isRequired,
  confirmationButtonText: PropTypes.string.isRequired,
  message: PropTypes.element.isRequired,
  title: PropTypes.string.isRequired,
  closeModal: PropTypes.func.isRequired,
  confirmModal: PropTypes.func.isRequired,
};
