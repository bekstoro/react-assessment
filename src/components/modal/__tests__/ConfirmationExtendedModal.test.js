import React from 'react';
import { render, within } from '@testing-library/react';

import ConfirmationExtendedModal from '@components/modal/ConfirmationExtendedModal';

describe('ConfirmationExtendedModal', () => {
  beforeEach(() => {
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'app');
    document.body.appendChild(modalRoot);
  });

  it('ConfirmationExtendedModal renders correctly', () => {
    render(
      <ConfirmationExtendedModal
        confirmModal={jest.fn()}
        confirmationButtonText=""
        cancellationButtonText=""
        closeModal={jest.fn()}
        isOpen
        message={<section />}
        title=""
      />
    );
    const { getByTestId } = within(document.getElementById('app'));
    expect(getByTestId('confirmation-extended-modal')).toMatchSnapshot();
  });

  it('ConfirmationExtendedModal should return empty div', () => {
    const { container } = render(
      <ConfirmationExtendedModal
        confirmModal={jest.fn()}
        confirmationButtonText=""
        cancellationButtonText=""
        closeModal={jest.fn()}
        isOpen={false}
        message={<section />}
        title=""
      />
    );
    expect(container).toMatchSnapshot();
  });
});
