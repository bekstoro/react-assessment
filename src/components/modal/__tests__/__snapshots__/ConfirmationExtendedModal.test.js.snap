// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationExtendedModal ConfirmationExtendedModal renders correctly 1`] = `
<div
  class="Modal__BaseModal-sc-1081dim-0 Modal__BigModal-sc-1081dim-1 cyuJJj eGHCUl"
  data-testid="confirmation-extended-modal"
>
  <h3
    class="Modal__ModalTitle-sc-1081dim-13 iVwSGl"
  >
    <i
      class="Modal__RedIcon-sc-1081dim-14 bVReHm material-icons"
    >
      warning
    </i>
    
  </h3>
  <section
    class="Modal__ModalContent-sc-1081dim-9 fQvezR"
  >
    <section />
  </section>
  <input
    class="Modal__ModalInput-sc-1081dim-11 krorRr"
    value=""
  />
  <section
    class="Modal__ModalButtons-sc-1081dim-6 fAOmbq"
  >
    <button
      class="Modal__ModalConfirmationButton-sc-1081dim-8 jJDWcp"
    />
    <button
      class="Modal__ModalCancellationButton-sc-1081dim-7 gufRcH"
    />
  </section>
</div>
`;

exports[`ConfirmationExtendedModal ConfirmationExtendedModal should return empty div 1`] = `<div />`;
